File: src/Scripting/Lua/chain_of_responsibility.lua
local visited = {}

local handlers = {
  {
    name = "faq",
    can_handle = function(amount) return amount <= 50 end
  },
  {
    name = "billing",
    can_handle = function(amount) return amount <= 500 end
  },
  {
    name = "escalation",
    can_handle = function(_) return true end
  }
}

local function handle(amount)
  for _, handler in ipairs(handlers) do
    table.insert(visited, handler.name)
    if handler.can_handle(amount) then
      return handler.name, string.format("refund(%d)", amount)
    end
  end
  error("unhandled request")
end

local handled, result = handle(250)
print(string.format("visited=%s;handled=%s;result=%s", table.concat(visited, ">"), handled, result))

File: src/Scripting/PHP/chain_of_responsibility.php
<?php

declare(strict_types=1);

abstract class RefundHandler
{
    private ?RefundHandler $next = null;

    public function setNext(RefundHandler $next): RefundHandler
    {
        $this->next = $next;
        return $next;
    }

    public function handle(int $amount, array &$visited): string
    {
        $visited[] = $this->name();
        if ($this->canHandle($amount)) {
            return $this->name();
        }
        if ($this->next === null) {
            throw new RuntimeException('No handler accepted the request.');
        }
        return $this->next->handle($amount, $visited);
    }

    abstract protected function name(): string;
    abstract protected function canHandle(int $amount): bool;
}

final class FaqHandler extends RefundHandler
{
    protected function name(): string { return 'faq'; }
    protected function canHandle(int $amount): bool { return $amount <= 50; }
}

final class BillingHandler extends RefundHandler
{
    protected function name(): string { return 'billing'; }
    protected function canHandle(int $amount): bool { return $amount <= 500; }
}

final class EscalationHandler extends RefundHandler
{
    protected function name(): string { return 'escalation'; }
    protected function canHandle(int $amount): bool { return true; }
}

$faq = new FaqHandler();
$billing = new BillingHandler();
$escalation = new EscalationHandler();
$faq->setNext($billing)->setNext($escalation);

$visited = [];
$handled = $faq->handle(250, $visited);
echo 'visited=' . implode('>', $visited) . ';handled=' . $handled . ';result=refund(250)' . PHP_EOL;
